### ISSUE TITLE
ValueError: Tag field could not prepare unexpected value

### ISSUE THREAD

**Issue by valentijnscholten** [radiac/django-tagulous#107]
Hi,

I am replacing `django-tagging` by `django-tagulous`. So far so good, the schema+data migration works fine after a little fiddling around. 

Next step is getting the forms to render correctly. Whatever I try, I keep running into the exception below. The form renders fine when there are no tags to show, or just 1. But as soon as there are 2 or more tags attached to the object I'm viewing, I am getting this exception. Seems to be related to the fact that the `TagField` is being asked to prepare a value which is a list of tags instead of a tag_string.

```
Traceback (most recent call last):
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/valentijn/dd/dojo/user/helper.py", line 77, in _wrapped
    return view_func(request, lookup_value, *args, **kwargs)
  File "/home/valentijn/dd/dojo/product/views.py", line 856, in edit_product
    return render(request,
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/shortcuts.py", line 36, in render
    content = loader.render_to_string(template_name, context, request, using=using)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/loader.py", line 62, in render_to_string
    return template.render(context, request)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/backends/django.py", line 61, in render
    return self.template.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 171, in render
    return self._render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/test/utils.py", line 96, in instrumented_test_render
    return self.nodelist.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/test/utils.py", line 96, in instrumented_test_render
    return self.nodelist.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 188, in render
    return template.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 173, in render
    return self._render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/test/utils.py", line 96, in instrumented_test_render
    return self.nodelist.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/defaulttags.py", line 209, in render
    nodelist.append(node.render_annotated(context))
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/defaulttags.py", line 309, in render
    return nodelist.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 993, in render
    return render_value_in_context(output, context)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/template/base.py", line 972, in render_value_in_context
    value = str(value)
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/utils/html.py", line 388, in <lambda>
    klass.__str__ = lambda self: mark_safe(klass_str(self))
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/forms/boundfield.py", line 33, in __str__
    return self.as_widget()
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/forms/boundfield.py", line 91, in as_widget
    value=self.value(),
  File "/home/valentijn/venv/lib/python3.8/site-packages/django/forms/boundfield.py", line 127, in value
    return self.field.prepare_value(data)
  File "/home/valentijn/venv/lib/python3.8/site-packages/tagulous/forms.py", line 339, in prepare_value
    return super(TagField, self).prepare_value(value)
  File "/home/valentijn/venv/lib/python3.8/site-packages/tagulous/forms.py", line 204, in prepare_value
    raise ValueError(_("Tag field could not prepare unexpected value"))
ValueError: Tag field could not prepare unexpected value
```

At https://github.com/radiac/django-tagulous/blob/b73b6c95cf07f28ee433e1b8eb7c560fc29e73f2/tagulous/forms.py#L201 there's:

```
    def prepare_value(self, value):
        print('prepare value for: ' + str(value))
        """
        Prepare the value - a tag string
        """
        # Catch no value (empty form for add)
        if not value:
            tag_string = ""

        # Will normally get a string
        elif isinstance(value, six.string_types):
            tag_string = value

        # Otherwise will be given by the model's TagField.value_from_object().
        # The value comes from django.forms.model.model_to_dict, which thinks
        # it produced a list of pks - but TagField.value_from_object tricked it
        # by returning a list with a single item, the tag string.
        else:
            # Catch changes in model_to_dict which broke the trick
            if len(value) != 1:
                raise ValueError(_("Tag field could not prepare unexpected value"))
            tag_string = value[0]

        return super(BaseTagField, self).prepare_value(tag_string)
```

When I debug I can see that `value` being sent in here is:

`['apple', 'beer', 'fries', 'vue']`

which causes the `ValueError` to be raised.

I think there's some code to prevent this from happening in `value_from_object`: https://github.com/radiac/django-tagulous/blob/b73b6c95cf07f28ee433e1b8eb7c560fc29e73f2/tagulous/models/fields.py#L475

I can see that that is working or at least triggered when debugging:

`['apple, beer, fries, vue']`

So I am not sure what's going wrong. I think my use case is pretty basic and standard, so it should just work?
Am I missing something?

I added a quick hack to be able to continue:
```
        # Otherwise will be given by the model's TagField.value_from_object().
        # The value comes from django.forms.model.model_to_dict, which thinks
        # it produced a list of pks - but TagField.value_from_object tricked it
        # by returning a list with a single item, the tag string.
        else:
            # Catch changes in model_to_dict which broke the trick
            if len(value) != 1:
                tag_string = render_tags(value)
            else:
                tag_string = value[0]
```

My model: 
https://github.com/valentijnscholten/django-DefectDojo/blob/f882f7f204ea0359bf7730d31798238b00134371/dojo/models.py#L669

```
from tagulous.models import TagField
class Product(models.Model):
    name = models.CharField(max_length=255, unique=True)
    description = models.CharField(max_length=4000)
    tags = TagField()
```

My form: 
https://github.com/valentijnscholten/django-DefectDojo/blob/f882f7f204ea0359bf7730d31798238b00134371/dojo/forms.py#L192
```
class ProductForm(forms.ModelForm):
    class Meta:
        model = Product
        fields = ['name', 'description', 'tags']
```

We render forms using: 
https://github.com/valentijnscholten/django-DefectDojo/blob/f882f7f204ea0359bf7730d31798238b00134371/dojo/templates/dojo/form_fields.html#L76
```
{% block css %}
    {{ form.media.css }}
{% endblock %}
{% block js %}
    {{ form.media.js }}
{% endblock %}
{% for field in form.visible_fields %}
            <div class="col-sm-10 {{ classes.value }} {% if field|is_multiple_checkbox %}multiple-checkbox{% endif %}">

                {{ field }}

                {% for error in field.errors %}
                    <span class="help-block {{ form.error_css_class }}">{{ error }}</span>
                {% endfor %}
            </div>
{% endfor %}
```

**Comment by valentijnscholten** [comment]
Form is called like this:
```
        form = ProductForm(instance=prod,
                           initial={'auth_users': prod.authorized_users.all()})
```
So no initial value for tags.

**Comment by valentijnscholten** [comment]
OK. Before posting I triple checked everything to be sure..... I removed the `initial` part where the tags where loaded as a list, problem was still happening. Turns out that a little later there is another statement that sets the initial value for tags... But in a different way, so was not caught in my global grep and search.... Anyway still might be good to consider supporting this scenario?
```
form.initial['tags'] = [tag.name for tag in prod.tags.all()]
```
